To read HTML webpages into pandas \
    - import pandas as pd \
    - use pd.read_html() function to read HTML webpage

In [64]:
import pandas as pd
url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
df= pd.read_html(url, header = 0)

In [65]:
df_canada=df[0]
df_canada.head()

Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

Only the cells that have an assigned borough are processed. So the cells with a borough that is 'Not assigned' will be dropped.

In [66]:
import numpy as np
df_canada["Borough"].replace('Not assigned', np.nan, inplace=True)
df_canada.dropna(subset=["Borough"], axis=0, inplace=True)
df_canada.reset_index(drop=True, inplace=True)
df_canada.head()

Postal Code           Borough                                Neighbourhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

More than one neighborhood can exist in one postal code area. These  rows will be combined into one row with the neighborhoods separated with a comma.

In [67]:
df_canada.shape

(103, 3)

In [68]:
df_canada["Postal Code"].value_counts

<bound method IndexOpsMixin.value_counts of 0      M3A
1      M4A
2      M5A
3      M6A
4      M7A
      ... 
98     M8X
99     M4Y
100    M7Y
101    M8Y
102    M8Z
Name: Postal Code, Length: 103, dtype: object>

The number of raws and length of value_counts of Postal Codes are equal(103). So there aren't any postal code area in which more than one neighbourhood exist.

If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough

In [69]:
df_canada["Neighbourhood"].replace('Not assigned', df_canada["Borough"], inplace=True)
df_canada

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [62]:
df_canada.shape

(103, 3)